### PR TITLE
split up cbmc in cbmc and jbmc

### PR DESCRIPTION
**Pull request by kroening**
The command-line syntax for C and Java is sufficiently different, which motivates the split-up into two binaries.

### CODE REVIEW

**Comment by tautschnig** [comment]
I think a few more regression test directories should be adapted:
```
regression$ find . -name "*.class" | cut -f2 -d/ | sort | uniq
cbmc
cbmc-cover
cbmc-java
cbmc-java-inheritance
goto-analyzer
strings
strings-smoke-tests
```

**Comment by lucasccordeiro** [comment]
I think this might also require some cleanup in the Java frontend (e.g., the input parameter `module` is not actually used in `bool java_bytecode_languaget::typecheck( symbol_tablet &symbol_table, const std::string &module)`).

**Comment by reuk** [comment]
I tried building this with CMake. There's a linking issue where `jbmc` can't find symbols from `cbmc_solvers` which is built into the `cbmc`module. I'm guessing `jbmc` shouldn't be dependent on `cbmc-lib`, so maybe `cbmc_solvers` should be moved to a different module, which both `jbmc` and `cbmc` can link against.

**Comment by reuk** [comment]
Having checked its makefile, it looks like `jbmc` includes several objects from the `cbmc` build. Maybe all the common components could go into a `bmc` module, which is included by both executables.

**Comment by kroening** [comment]
This is known to disrupt anyone using cbmc for Java; it is hoped that users of cbmc for C will not notice this PR.

However, going forward, there will be less disruption of either the Java or C side by changes to the respective other side.

File: src/cbmc/cbmc_parse_options.cpp [hunk 1]
```diff
@@ -757,12 +745,10 @@ bool cbmc_parse_optionst::process_goto_program(
       get_message_handler(),
       goto_model,
       cmdline.isset("pointer-check"));
-    // Java virtual functions -> explicit dispatch tables:
+    // virtual functions -> explicit dispatch tables:
     remove_virtual_functions(goto_model);
     // remove catch and throw (introduces instanceof)
     remove_exceptions(goto_model);
```
Comment (smowton) [inline src/cbmc/cbmc_parse_options.cpp]: This introduces java_instanceof, so you should have either this and remove_instanceof or neither

File: src/cbmc/cbmc_parse_options.cpp [hunk 2]
```diff
@@ -873,7 +835,6 @@ bool cbmc_parse_optionst::process_goto_program(
 
     // remove any skips introduced since coverage instrumentation
     remove_skip(goto_model);
-    goto_model.goto_functions.update();
```
Comment (smowton) [inline src/cbmc/cbmc_parse_options.cpp:876]: Why remove this?

File: src/cbmc/cbmc_parse_options.h [hunk 3]
```diff
@@ -67,8 +65,6 @@ class optionst;
   "(string-abstraction)(no-arch)(arch):" \
   "(round-to-nearest)(round-to-plus-inf)(round-to-minus-inf)(round-to-zero)" \
   "(graphml-witness):" \
-  JAVA_BYTECODE_LANGUAGE_OPTIONS \
-  "(java-unwind-enum-static)" \
```
Comment (smowton) [inline src/cbmc/cbmc_parse_options.h:71]: The removals here and in the help text don't match -- suggest removing `main-class` and `classpath` as well.

File: src/jbmc/jbmc_parse_options.cpp [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+
+  //
+  // Print a banner
+  //
+  status() << "JBMC version " CBMC_VERSION " "
+           << sizeof(void *)*8 << "-bit "
+           << config.this_architecture() << " "
+           << config.this_operating_system() << eom;
+
+  register_language(new_ansi_c_language);
```
Comment (smowton) [inline src/jbmc/jbmc_parse_options.cpp:432]: Why register C? If it's purely for a fallback for `from_expr` et al suggest comment to this effect.

**Review by smowton — COMMENTED** [review]
Some specific fixes I think ought to be made.

General notes:

1. Previously Java changes were breaking C, and clearly this was a problem. However this exchanges that problem for around a thousand lines of duplicated code between cbmc and jbmc which will accidentally diverge and cause the opposite problem. Could we do better by restructuring `process_goto_functions`, the main source of trouble, to use post-pass structures that explicitly state their language applicability, such that e.g. the convert_nondet mistake should not be repeated?

2. What about goto-analyzer and similar frontends that are still unified? It seems like there should be coto-analyzer and joto-analyzer, but this just deepens the duplication problem.

3. It's a small shame that this sort of split moves us further away from cross-language analysis, e.g. JNI.

If you're willing I'd like to propose an alternative PR that provides hooks for language-specific post-passes and thus makes language-independent frontends safer.

File: src/jbmc/jbmc_parse_options.cpp (hunk 4 above)
Comment (kroening) [reply src/jbmc/jbmc_parse_options.cpp:432]: No, we actually do add C-mode symbols (rounding mode, etc.)

File: src/cbmc/cbmc_parse_options.cpp (hunk 2 above)
Comment (kroening) [reply src/cbmc/cbmc_parse_options.cpp:876]: remove_skip now does it.

File: src/cbmc/cbmc_parse_options.cpp (hunk 1 above)
Comment (kroening) [reply src/cbmc/cbmc_parse_options.cpp]: Done

File: src/cbmc/cbmc_parse_options.h (hunk 3 above)
Comment (kroening) [reply src/cbmc/cbmc_parse_options.h:71]: Done.

**Comment by kroening** [comment]
General comment: This is mostly about the command-line interface, which is not necessarily desirable to be uniform. I'd be happy about a proposal that removes duplication on the goto-program processing.

**Comment by kroening** [comment]
BTW, solver selection appears to be another thing that could be de-duplicated.

**Pull request merged**